---
title: "July 18 exploration"
output: html_notebook
---

# Load Packages
```{r}
source("/beegfs/ay1392/TET2/scRNAseq/Analysis/scRNAscripts/R/hashtags_create.R")
source("/beegfs/ay1392/TET2/scRNAseq/Analysis/scRNAscripts/R/module_scores.R")
source("/beegfs/ay1392/TET2/scRNAseq/Analysis/scRNAscripts/R/cluster.R")

suppressPackageStartupMessages({
  library(magrittr)
  library(glue)
  library(Seurat)
  library(future)
  library(Matrix)
  library(tidyverse)
  library(data.table)
  library(cowplot)
  library(scales)
  library(pheatmap)
  library(RColorBrewer)
  library(ggsci)
  library(eulerr)
  library(UpSetR)
  library(GGally)
  library(compositions)
})

```



## Cluster non var genes

```{r}
# re integrate with 4,000 genes

# pca, umap, tsne with botton 2,000

# cluster

```



## Cluster Passegue and Ido Amit genes

```{r}
# reintegrate with 2,000 genes plus passegue and Ido Amit

# pca, umap, tsne

# cluster

```


## Get cluster stats 
```{r}
seurat_obj <- readRDS("/beegfs/ay1392/TET2/scRNAseq/Analysis/re-do/integrate_all2/seurat_obj.rds")

s_obj =
    FindNeighbors(
      seurat_obj, dims = 1:30, k.param = 30,
      graph.name = "snn", compute.SNN = TRUE, force.recalc = TRUE
    )
s_obj = FindClusters(s_obj, algorithm = 3, resolution = 5, graph.name = "snn", verbose = FALSE)

Idents(s_obj) <- "snn_res.5"
label <- "snn_res5"
# snn_res5
calculate_cluster_stats(s_obj, label)

counts_csv <- "/home/ay1392/anna_beegfs/TET2/scRNAseq/Analysis/re-do/integrate_all2/expression.mean.snn_res5.csv"
cluster_counts <- data.table::fread(file = counts_csv, showProgress = FALSE, data.table = FALSE, nThread = 4)
colnames(cluster_counts) <- cluster_counts[1,]
cluster_counts <- cluster_counts[-1,]
rownames(cluster_counts) <- cluster_counts$gene
cluster_counts <- cluster_counts[,-1]
```


## Cell Cycle

```{r}
# read in raw counts from snn_res5 clust averages
counts_raw <- cluster_counts
  
# read in metadata for the integrated set
metadata_csv <- "/home/ay1392/anna_beegfs/TET2/scRNAseq/Analysis/re-do/integrate_all2/metadata.snn_res5.csv"
meta = data.table::fread(file = metadata_csv, showProgress = FALSE, data.table = FALSE, nThread = 4)

# module score cell cycle
cc <- cc.genes # cell cycle genes from seurat
cc <- as.data.frame(t(as.data.frame(transpose(cc))))
colnames(cc) <- c("s.genes", "g2m.genes")
module_tbl <- cc %>% 
  gather(na.rm = TRUE) %>% 
  rename(celltype = key) %>% 
  rename(gene = value)

cellcycle_mod_score <- module_score(module_tbl = module_tbl, 
                                    counts_norm = NULL,
                                    counts_raw = NULL, 
                                    method = "rsscore")           
# merge metadata with cellcycle mod score

# 


# pca, umap, tsne
```

## Regress out Cell Cycle
```{r}
s_obj <- CellCycleScoring(s_obj, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# regress out cell cycle
s_obj$CC.Difference <- s_obj$S.Score - s_obj$G2M.Score
s_obj <- ScaleData(s_obj, vars.to.regress = "CC.Difference", features = rownames(s_obj))
# module score cell cycle

# pca, umap, tsne
```


## Gene sets
```{r}
# msigdbr


#Ido Amit (fig2 and fig3)
fig2 <- c("Ifitm1", 
         "Ifitm3",
         "Txnip",
         "Pf4",
         "Cd34",
         "Flt3",
         "Gata1",
         "Car2", 
         "Mt2",
         "Car1",
         "Hba-a2",
         "Hba-b1",
         "Fcgr3",
         "Ctsg",
         "Mpo",
         "Ly6c2",
         "Ly86",
         "F13a1",
         "Csf1r",
         "Lyz1",
         "Lyz2",
         "Elane",
         "Fcnb",
         "Gstm1",
         "Camp",
         "Ltf",
         "Cd63",
         "Prss34",
         "Prg2",
         "Cd74",
         "H2-Aa",
         "Ly6d",
         "Siglech",
         "Dntt",
         "Vpreb1",
         "Vpreb3",
         "Cd79b",
         "Fcrla",
         "Ccl5",
         "Nkg7",
         "C1qb")

#GATA1 inflammosome


```


## Lineage Bias in WT 
```{r}
# Ido amit and Passegue trends in WT

# TFs from G and A trends in WT

```


## Lineage Bias in HR
```{r}
# Ido amit and Passegue trends in HR

# TFs from G and A trends in HR

```


## Splice variants of TET2
```{r}

```

